### PR TITLE
Feature/transaction per project more db connection options standalone importer.

### PR DESCRIPTION
**Pull request by dvanderleij**
Hi all,

Here are some quality of life changes. i will push my test improvements in a different PR to keep them somewhat separated.

### CODE REVIEW

File: src/main/java/refactoringml/ProcessMetricsCollector.java [hunk 1]
```diff
@@ -164,8 +190,11 @@ private void outputNonRefactoredClass (ProcessMetricTracker pmTracker) throws IO
 			// ... as well as in the temp one, so that we can calculate the CK metrics
 			writeFile(tempDir + pmTracker.getFileName(), sourceCodeBackThen);
 
-			if(pmTracker.getBaseCommitMetaData().getId() == 0)
-				db.persist(pmTracker.getBaseCommitMetaData());
+			if(pmTracker.getBaseCommitMetaData().getId() == 0) {
+				CommitMetaData toSave = pmTracker.getBaseCommitMetaData();
+				db.save(toSave);
+				toRemoveOnException.add(toSave);
```
Comment (jan-gerling) [inline src/main/java/refactoringml/ProcessMetricsCollector.java:196]: Did you test this? Are you sure, there are no cases in which this is not causing database errors, e.g. due to FOREIGN KEY Constraints?

File: src/main/java/refactoringml/db/Database.java [hunk 2]
```diff
@@ -1,144 +1,106 @@
...
+	public void save(Object obj) {
+		session.save(obj);
 	}
 
 	public void update(Object obj) {
 		session.update(obj);
 	}
 
-	public void close() {
+	public void deleteWithoutThrowingOnException(Object obj) {
```
Comment (jan-gerling) [inline src/main/java/refactoringml/db/Database.java:38]: Did you test this? Are you sure, there are no cases in which this is not causing database errors, e.g. due to FOREIGN KEY Constraints?

File: src/main/java/refactoringml/RunImportCsv.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+		@Parameter(names = { "-t",
+				"thread-amount" }, description = "Amount of threads to run in parallel. Will default to amount of cores available")
+		private int runWithCores = Runtime.getRuntime().availableProcessors();
+
+		@Parameter(required = true, description = "Input csv file")
+		private String inputCsvFile;
+
+	}
+
+	private void run(Args args, String repoLocation) throws InterruptedException, IOException {
```
Comment (jan-gerling) [inline src/main/java/refactoringml/RunImportCsv.java:89]: How does this behave under high load? E.g. huge projects with enormous files?

File: src/main/java/refactoringml/RunImportCsv.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+		ThreadPoolExecutor tp = (ThreadPoolExecutor) Executors.newFixedThreadPool(args.runWithCores);
+		tp.setCorePoolSize(args.runWithCores);
+		SessionFactory sf = HibernateConfig.getSessionFactory(args.databaseHost, Optional.ofNullable(args.databasePort),
+				args.databaseName, args.databaseUser, args.databasePassword, args.hbm2ddlStrategy,
+				DatabaseEngine.valueOf(args.databaseEngine), ConnectionPoolType.valueOf(args.connectionPoolType));
+		try {
+			try (Stream<String> stream = Files.lines(Paths.get(args.inputCsvFile))) {
+				List<RepoProcesser> tasks = stream.map(line -> new RepoProcesser(repoLocation, args.storeSourceCode,
+						new Database(sf.openSession()), line)).collect(Collectors.toList());
+				tp.invokeAll(tasks, 100, TimeUnit.HOURS);
```
Comment (jan-gerling) [inline src/main/java/refactoringml/RunImportCsv.java:100]: What happens here in case of an Error, e.g. OutOfMemoryError?

File: src/main/java/refactoringml/RunImportCsv.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
+package refactoringml;
```
Comment (jan-gerling) [inline src/main/java/refactoringml/RunImportCsv.java:1]: Can you please transfer this into a new PR? It is unrelated to the other proposed changes? And I find it difficult to assess.

File: src/main/java/refactoringml/db/Database.java [hunk 6]
```diff
@@ -1,144 +1,106 @@
 package refactoringml.db;
```
Comment (jan-gerling) [inline src/main/java/refactoringml/db/Database.java:1]: Can you please transfer the changes to the Database connection and exception handling into a new PR, because it is unrelated to the other quality of life changes. Also, I am not certain how some of the changes behave in specific situations, e.g. store CommitMetaData in db -> store related RefactoringInstance in db -> Exception -> delete CommitMetaData from db -> Foreign Key Constraint Error on db?

File: src/main/java/refactoringml/db/HibernateConfig.java [hunk 7]
```diff
@@ -1,89 +1,175 @@
...
-import java.util.Properties;
-
-public class HibernateConfig {
+public abstract class HibernateConfig {
 
-	private static SessionFactory sessionFactory;
+	protected final Configuration configuration = new Configuration();
 
-	public SessionFactory getSessionFactory(String url, String user, String pwd) {
+	public static SessionFactory getSessionFactory(String url, String user, String pwd) {
```
Comment (jan-gerling) [inline src/main/java/refactoringml/db/HibernateConfig.java:18]: In the old implementation, the URL for the DB could also include additional information e.g. the port, usage of ssl for authenticiation, DB time zone etc. How do you handle it in this implementation?

File: src/main/java/refactoringml/RefactoringAnalyzer.java [hunk 8]
```diff
@@ -220,47 +298,44 @@ private RefactoringCommit calculateCkMetrics(ImmutablePair<String,String> refact
...
+					refactoringErrorState);
 		} else {
 			for (RefactoringCommit refactoringCommit : refactorings) {
-				db.persist(refactoringCommit);
+				db.save(refactoringCommit);
 			}
 		}
 
-		return refactorings.isEmpty()? null : refactorings.get(0);
+		return refactorings;
```
Comment (jan-gerling) [inline src/main/java/refactoringml/RefactoringAnalyzer.java:339]: Why move the entire, potentially empty, list to collectCommitData()?

File: src/main/java/refactoringml/RefactoringAnalyzer.java [hunk 9]
```diff
@@ -42,102 +76,129 @@ public RefactoringAnalyzer (Project project, Database db, Repository repository,
...
+								refactoredClassName, refactoring.getName(), commit.getName());
 					}
 				}
 			}
-		} catch (Exception e){
-			log.error("Failed to collect commit data for a refactored commit." + createErrorState(superCommitMetaData.getCommitId(), project), e);
+		} catch (Exception e) {
+			log.error("Failed to collect commit data for a refactored commit. {}",
+					createErrorState(superCommitMetaData.getCommitId(), project), e);
+			toRemoveOnException.forEach(db::deleteWithoutThrowingOnException);
```
Comment (jan-gerling) [inline src/main/java/refactoringml/RefactoringAnalyzer.java:170]: Why did you add this list(toRemoveOnException)? In case of an exception, we are rerolling the last transaction (the last commit) in App.java anyways.

File: src/main/java/refactoringml/RefactoringAnalyzer.java [hunk 10]
```diff
@@ -220,47 +298,44 @@ private RefactoringCommit calculateCkMetrics(ImmutablePair<String,String> refact
...
-			log.error("CK did not find class " + refactoredClasses.getLeft() + "/" + refactoredClasses.getRight() + createRefactoringErrorState(commitMetaData.getCommitId(), project, refactoringSummary));
+		if (refactorings.isEmpty()) {
+			String refactoringErrorState = createRefactoringErrorState(commitMetaData.getCommitId(), project,
+					refactoringSummary);
+			log.error("CK did not find class {}/{} {}", refactoredClasses.getLeft(), refactoredClasses.getRight(),
+					refactoringErrorState);
 		} else {
 			for (RefactoringCommit refactoringCommit : refactorings) {
-				db.persist(refactoringCommit);
+				db.save(refactoringCommit);
```
Comment (jan-gerling) [inline src/main/java/refactoringml/RefactoringAnalyzer.java:335]: This applies also to the other uses of save instead of persist.

Why would you use a save here? In my opinion, persist is the saver option:
1. it guarantees that it is not inserted outside a transaction, thus we can reroll the last transaction if necessary. Or in cases of crashes never commit the transaction and thereby, keep the DB cleaner.
2. The id created for the object is not created when it is called, but when it is flushed. Thus, we either have no race conditions between multiple workers trying to assign the same id to an object or the database needs less locks.

File: src/main/java/refactoringml/App.java [hunk 11]
```diff
@@ -108,173 +118,203 @@ public App (String datasetName,
...
 			long startTimeTransaction = System.currentTimeMillis();
-			db.commit();
-			log.debug("Committing the transaction for commit " + commitHash + " took " + (System.currentTimeMillis() - startTimeTransaction) + " milliseconds.");
+			log.debug("Committing the transaction for commit {} took {} milliseconds.", commitHash,
+					(System.currentTimeMillis() - startTimeTransaction));
 		} catch (Exception e) {
+			// are not supported by mariadb
 			exceptionsCount++;
-			log.error("Unhandled exception when collecting commit data for commit: " + commitHash + createErrorState(commitHash, project), e);
-			db.rollback(createErrorState(commitHash, project));
```
Comment (jan-gerling) [inline src/main/java/refactoringml/App.java:259]: Why did you remove the transaction rollback?
In my eyes it is a reliable and save way to remove the last commit from the DB. Which might not even be the best idea at all, as we lose some data points, that we could keep and clear up the DB later.

**Review by jan-gerling — CHANGES_REQUESTED** [review]

**Comment by jan-gerling** [comment]
> Here are some quality of life changes. i will push my test improvements in a different PR to keep them somewhat separated.

Thanks for the PR, I have seen some good changes.
I have a some remarks though:
1. Could you please split this huge PR into multiple smaller ones, that are more concise? I find it impossible to assess the impacts of all changes together. Also for some of the changes, I am not sure how the would perform under high load or some other edge cases.
2. Some of the suggested changes I could not understand. I made comments for the most important ones I noticed.

File: src/main/java/refactoringml/RefactoringAnalyzer.java (hunk 10 above)
Comment (dvanderleij) [reply src/main/java/refactoringml/RefactoringAnalyzer.java:335]: I thought save was just a hibernate specific way of persist and since we are using hibernate specific sessions and sessionfactories rather than entitymanagers and entitymanagerfactories I figured it was a good idea to be consistent and use hibernate specific methods as well. After looking it up it seems that it does not just have extra functionality of returning an id flushing and detaching the object. I will change this back to persist. Maybe it is an idea to generify the session to entitymanager though since we are mostly only using entitymanager methods anyway.

File: src/main/java/refactoringml/db/HibernateConfig.java (hunk 7 above)
Comment (dvanderleij) [reply src/main/java/refactoringml/db/HibernateConfig.java:18]: I made sure to leave the old API intact so calling the service would be possible in the original way and added functionality if wanted for connectionpools etc. The url is passed as is and the tests use some of these options you mention

File: src/main/java/refactoringml/App.java (hunk 11 above)
Comment (dvanderleij) [reply src/main/java/refactoringml/App.java:259]: I wanted to encapsulate a single run for a project in a single transaction. So that if an error outside the commit analysis occurs there is no data related to that project is left over in the db. Unfortunately most open source db engines don't support nested transactions as far as i know. I came up with a solution that keeps track of the persisted entities during the commit analysis and removes them in the session if any errors occur. However in hindsight it seems that ll persists are of the commits with the lone exception of saving the project. So the previous solution suffices.

File: src/main/java/refactoringml/RunImportCsv.java (hunk 3 above)
Comment (dvanderleij) [reply src/main/java/refactoringml/RunImportCsv.java:89]: Could you give me an example of a project with such files?. Then i can test it. I tested the first 1000 projects in projects_final.csv and there seemed to be no issues.

File: src/main/java/refactoringml/RunImportCsv.java (hunk 4 above)
Comment (dvanderleij) [reply src/main/java/refactoringml/RunImportCsv.java:100]: I have not encountered it as of yet. This class was made with the idea for collecting data on smaller datasets without having to start a lot of infrastructure.

**Comment by dvanderleij** [comment]
Sure i'll create a PR for every commit there

File: src/main/java/refactoringml/db/Database.java (hunk 2 above)
Comment (dvanderleij) [reply src/main/java/refactoringml/db/Database.java:38]: See my comment regarding the transaction below

File: src/main/java/refactoringml/ProcessMetricsCollector.java (hunk 1 above)
Comment (dvanderleij) [reply src/main/java/refactoringml/ProcessMetricsCollector.java:196]: See my comment regarding the transaction below

**Comment by dvanderleij** [comment]
See #22 #23 #25 #24 #26 for the separate pr's they are numbered to be reviewed in order
